### PR TITLE
New credit card workflow - TPWDCEE-3489

### PR DESCRIPTION
**Pull request by pretzelhands**

### CODE REVIEW

File: wirecardpaymentgateway/helper/TransactionBuilder.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    /**
+     * Constructs the requested transaction for the payment type
+     *
+     * @return \Wirecard\PaymentSdk\Transaction\Transaction
+     * @throws \Exception
+     * @since 1.4.0
+     */
+    public function buildTransaction()
+    {
+        if (!isset($this->orderId)) {
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TransactionBuilder.php]: i woud write only $transaction->setEntity() here and then for each value a private function for example.
`
$transaction->setRedirect($this->getRedirects());

private function getRedirects() {
    ...
}
`

File: wirecardpaymentgateway/helper/TransactionBuilder.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            $this->module->createRedirectUrl($this->orderId, $this->paymentType, 'failure')
+        );
+
+        /** @var Transaction $transaction */
+        $transaction = $payment->createTransaction($this->module, $this->cart, \Tools::getAllValues(), $this->orderId);
+        $transaction->setNotificationUrl($this->module->createNotificationUrl($cartId, $this->paymentType));
+        $transaction->setRedirect($redirectUrls);
+        $transaction->setAmount(new Amount((float)$amount, $currency->iso_code));
+
+        $customFields = new CustomFieldCollection();
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TransactionBuilder.php]: addCustomField method

File: wirecardpaymentgateway/helper/TransactionBuilder.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        $customFields->add(new CustomField('orderId', $this->orderId));
+        $transaction->setCustomFields($customFields);
+
+
+        if (\Tools::getValue('token_id') && $transaction instanceof CreditCardTransaction) {
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TransactionBuilder.php]: I would first check for the right transaction type and then for the token. As i save some time like that. Also i woud make an isToken() out of it.

File: wirecardpaymentgateway/helper/TransactionBuilder.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+     * @since 1.4.0
+     */
+    public function __construct($module, $context, $cartId, $paymentType)
+    {
+        $this->module = $module;
+        $this->context = $context;
+        $this->cart = new \Cart((int) $cartId);
+        $this->paymentType = $paymentType;
+
+        $this->additionalInformation = new AdditionalInformation();
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TransactionBuilder.php]: For me it is an additionalInformationBuilder consider renaming it

File: wirecardpaymentgateway/helper/TransactionBuilder.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        $customFields = new CustomFieldCollection();
+        $customFields->add(new CustomField('orderId', $this->orderId));
+        $transaction->setCustomFields($customFields);
+
+
+        if (\Tools::getValue('token_id') && $transaction instanceof CreditCardTransaction) {
+            $transaction->setTokenId(\Tools::getValue('token_id'));
+        }
+
+        if ($this->module->getConfigValue($this->paymentType, 'shopping_basket')) {
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TransactionBuilder.php]: i would do methods as addBasket() and then check inside it if i need to add it then im left with a clean method thats readable without the getConfig stuff. Same for descriptor and additional fields...

File: wirecardpaymentgateway/models/Payment.php [hunk 6]
```diff
@@ -373,6 +378,15 @@ public function isAvailable($module, $cart)
         return true;
     }
 
+    public function getOperationForPaymentAction($action)
```
Comment (tomazpu) [inline wirecardpaymentgateway/models/Payment.php]: phpdoc missing

File: wirecardpaymentgateway/models/PaymentUnionPayInternational.php [hunk 7]
```diff
@@ -38,7 +38,7 @@
 use Wirecard\PaymentSdk\TransactionService;
```
Comment (tomazpu) [inline wirecardpaymentgateway/models/PaymentUnionPayInternational.php]: can UnionPayInternational just extend CreditCard and override the methods that are needed to be overridden?

File: wirecardpaymentgateway/controllers/front/configprovider.php [hunk 8]
```diff
@@ -54,8 +54,9 @@ public function initContent()
      */
     public function displayAjaxGetCreditCardConfig()
     {
+        $cartId = Tools::getValue('id_cart');
```
[the remark this answers is not in the document]
Comment (pretzelhands) [reply wirecardpaymentgateway/controllers/front/configprovider.php]: Send JavaScript param (e.g. `paymentType`?)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 9]
```diff
@@ -63,79 +65,44 @@ class WirecardPaymentGatewayPaymentModuleFrontController extends ModuleFrontCont
...
-        $additionalInformation = new AdditionalInformation();
-
-        if ($cart->id_customer == 0 || $cart->id_address_delivery == 0 || $cart->id_address_invoice == 0 ||
-            !$this->module->active
+        $existingOrderId = \Tools::getValue('order_number');
+        $cart = $existingOrderId
+            ? \Cart::getCartByOrderId($existingOrderId)
+            : $this->context->cart;
+
+        if ($cart->id_customer == 0
```
[the remark this answers is not in the document]
Comment (pretzelhands) [reply wirecardpaymentgateway/controllers/front/payment.php]: Same solution as above

**Pull request closed without merging**